### ISSUE TITLE
Merlin process opens short-lived cmd windows on Windows Emacs

### ISSUE THREAD

**Issue by leviroth** [ocaml/merlin#714]
To reproduce:

1. Install Ocaml on Windows 10 via https://fdopen.github.io/opam-repository-mingw/. I'm using version 4.04.0+mingw64.
   
2. Install merlin via opam. I am currently pinned to 8ea5c48d, but also saw this issue on 2.5.5.

2. Install Emacs via https://sourceforge.net/projects/emacsbinw64/. (I didn't test with other builds.) Version: GNU Emacs 25.1.1 (x86_64-w64-mingw32) of 2016-09-21

3. Add merlin to .emacs by copying the code at https://github.com/ocaml/merlin/wiki/emacs-from-scratch#installing-merlin-and-its-emacs-mode-with-opam.
   
3. Open Emacs via [ocaml-env-win](https://fdopen.github.io/opam-repository-mingw/ocaml-env/).
   
4. Open an Ocaml buffer and save. A cmd window will open very briefly any time merlin is invoked.
   
In contrast, ocp-indent works without a hitch, so this seems to be a problem with Merlin specifically.

The actual checking behavior works as expected, which is great. The cmd window a) is a visual annoyance and b) captures keystrokes, since it pops up during completion.

**Comment by leviroth** [comment]
After investigating, it looks like this is caused by the server invocation [here](https://github.com/ocaml/merlin/blob/d301f4971ae8cdeb7345db585283d3cfaa58aae2/emacs/merlin.el#L487). If I change `"server"` to `"single"`, the problem goes away. Is it possible that the server is opening a new command window and then exiting immediately?

**Comment by let-def** [comment]
I don't know why windows would open a command window (wrong compilation flags?!), @dra27 can maybe help? :)

**Comment by bryphe** [comment]
@jordwalke and I were spending time debugging this - I used [Process Monitor](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon) to grab a trace.

I found a few clues:
- This only reproduces when I am using `ppx`. For example, the [`hello-reason`](https://github.com/esy-ocaml/hello-reason) project does not exhibit this behavior, but the [`merlin-language-server`](https://github.com/bryphe/merlin-language-server) project does. (`hello-reason` uses no ppx, while `merlin-language-server` does).
- With the process monitor trace, I see calls to `ppx.exe` spawned using `cmd.exe /c`. The problematic process creations occurred in this form:

```
C:\Windows\system32\cmd.exe /c C:.\merlin-language-server_build\default\.ppx\0ba27ff973ee615f2faec1d1e9a8c7a3\ppx.exe --as-ppx --cookie "library-name=\"Protocol\"" "C:\Users\bryan\AppData\Local\Temp\camlppxcc291b" "C:\Users\bryan\AppData\Local\Temp\camlppx3a52ba" 1> NUL 2> NUL
```

I suspect that we could switch the `ppx.exe` process creation to a similiar strategy we use for the extenders - `Unix.create_process` to solve this. It seems the `cmd.exe /c` execution is spinning up these rogue short-lived command windows.

**Comment by bryphe** [comment]
I suspect it is related to the `Sys.command` here:
https://github.com/ocaml/merlin/blob/8afde7e89260e020ed372d67e184a8d15e1d8276/src/ocaml/driver/pparse.ml#L57

Would it be possible to use `Unix.create_process` here, instead, as we do for the extensions?

**Comment by jordwalke** [comment]
@bryphe: @dra27 suggested that we might need to do what this portion of ocamlmerlin.c does, if `Unix.create_process` isn't sufficient:

https://github.com/ocaml/merlin/blame/master/src/frontend/ocamlmerlin.c#L292

**Comment by bryphe** [comment]
Thanks @jordwalke @dra27 ! I'll test this out later today.

I suspect that `Unix.create_process` should be better behaved - as it seems like the logic is correct here: https://github.com/ocaml/ocaml/blob/3735b4ab20c15c01fe2a6c58138b354a354c8c21/otherlibs/win32unix/createprocess.c#L64
(It either reuses the existing console, or it uses a new one _but_ uses the `SW_HIDE` flag).

**Comment by bryphe** [comment]
I was able to hack together a change that works now (and `ppx` is fully recognized by merlin in my windows project - huge help for `ppx_left` and `ppx_pipebang`!)

I had to make a few changes -

I replaced the call to `Sys.command` with this:
```
let run_ppx_command cmd = 
    let pid = 
        Unix.create_process "cmd.exe" [|"/c"; cmd|] 
        Unix.stdin Unix.stderr Unix.stderr
    in
    let (_, exit) = 
        Unix.waitpid [] pid
    in
    match exit with
    | Unix.WEXITED x -> x
    | Unix.WSIGNALED x -> x
    | Unix.WSTOPPED x -> x
```

Under the hood, this is still essentially making the same call, but it takes advantage of the fact that `Unix.create_process` sets the `SW_HIDE` flag.

- I had to fix the path slashes in my `.merlin` file. I guess it is `dune` that generates this - so this would really be a bug in `dune`? Or do you think `merlin` should be able to handle mixed-slash paths, @dra27 ?

- I had to remove the _escaped quotes_. The command line being sent to this new command was:
```
E:/merlin-language-server/_build/default/.ppx/de456a5f258388d2d4b604b5a1dae6c7/ppx.exe  --as-ppx --cookie \"library-name=\\\"Server\\\"\" \"C:\Users\bryph\AppData\Local\Temp\camlppx25afa6 \" \"C:\Users\bryph\AppData\Local\Temp\camlppx370d12\"
```

But the extra `\` character caused problems. If I removed them, such that the command line is: 
```
E:/merlin-language-server/_build/default/.ppx/de456a5f258388d2d4b604b5a1dae6c7/ppx.exe  --as-ppx --cookie library-name=Server C:\Users\bryph\AppData\Local\Temp\camlppx25afa6  C:\Users\bryph\AppData\Local\Temp\camlppx370d12
```
(I did this by both removing the `Filename.quote` methods for input / output files, and manually editing my `.merlin` file)

**Comment by bryphe** [comment]
So remaining issues to figure out:
- What fixes need to go into the `.merlin` file generation in `dune`? Potentially, we could fix both the quote-escaping for the `--cookie` parameter and use a consistent path slashing there, to fix up the `merlin` files.
- What fixes should be done here in `merlin`? On Windows, we can switch to use that `Unix.create_process` variant, and _not_ quote the input/output files with `Filename.quote`. (We should still quote the paths - but not with the extra escape character)

**Comment by dra27** [comment]
This is great, thanks @bryphe! Are you doing this on Windows 10? I hope that the shenanigans with `SW_HIDE` still causing a flashing console window is just me vaguely remembering an older version of Windows misbehaving (or just me getting old.....)

The Dune issue is https://github.com/ocaml/dune/issues/366, I think. Merlin is behaving correctly regarding the escaping - the backslashes should be escaped and the problem should be fixed in Dune.

The escaping is a bigger problem - switching to `Unix.create_process` should allow `cmd /c ` to be avoided completely - but that involves reworking Merlin slightly to keep its command line arguments as lists for as long as possible. In the long-run this is worth changing - cmd escaping is just too weird.

**Comment by let-def** [comment]
If necessary we can add a windows-specific C binding for spawning the process in the right way.

**Comment by bryphe** [comment]
Thanks @dra27 and @let-def for the response!

> Are you doing this on Windows 10? I hope that the shenanigans with SW_HIDE still causing a flashing console window is just me vaguely remembering an older version of Windows misbehaving (or just me getting old.....)

Yes, I'm testing on Windows 10. It's certainly possible that `SW_HIDE` could be misbehaving on earlier versions. I thought it could also be due to the OCaml version... but looks like that `SW_HIDE` flag was added a long time ago 😄 I believe there are also cases where you can start a process with `SW_HIDE`, but that process can start a new process that still opens a console window... 

But I can say that using `Unix.create_process` fixes this particular window opening for me on Windows 10!

> The Dune issue is ocaml/dune#366, I think. Merlin is behaving correctly regarding the escaping - the backslashes should be escaped and the problem should be fixed in Dune.

Thank you for linking that issue! I'll investigate that next - it seems like we will need that fix in any case.

> The escaping is a bigger problem - switching to Unix.create_process should allow cmd /c to be avoided completely - but that involves reworking Merlin slightly to keep its command line arguments as lists for as long as possible. In the long-run this is worth changing - cmd escaping is just too weird.

Definitely! I only used `cmd.exe /c` above because of the funky format of the command. If we can switch to a format where we have a list, we can call the executable directly and pass those arguments in. And even with my change above - I _still_ had to modify the `.merlin` manually and remove some of the escaping for it work correctly.

@dra27 - regarding keeping command line arguments as lists - are you picturing that impacting the `.merlin` file, such that it records the arguments as a list, like a sexp? ie:

__BEFORE:__ `FLG -ppx "E:\merlin-language-server\_build/default/.ppx/0ba27ff973ee615f2faec1d1e9a8c7a3/ppx.exe --as-ppx --cookie \"library-name=\\\"Protocol\\\"\""`
__AFTER:__ `FLG -ppx ("E:\merlin-language-server\_build/default/.ppx/0ba27ff973ee615f2faec1d1e9a8c7a3/ppx.exe" "--as-ppx" "--cookie \"library-name=\"Protocol\"")`

Or did you have another way of accomplishing this in mind?

The next steps for my investigation are:
- Propose a PR here to use `Unix.create_process` on Windows (fixes the flashing window, at least)
- Investigate dune#66 to fix the initial path escaping issue on Windows (fixes the initial error where the `ppx.exe` is not found)

However, we still have the remaining issue of `cmd /c` avoidance - if we plan on changing the `.merlin` format as described above, that would involve changes to both `dune` and `merlin`. Is there a precedent for making these changes in a backward-compatible way? The scenario I worry about is if someone upgrades `dune` but not `merlin` or vice-versa. Or is there another sort of fix that doesn't involve changing the `.merlin` format as aggressively?

Thanks again for your help!

**Comment by jordwalke** [comment]
Looks like Dune has fixed their path constructions. I have tested and they use either `/` or `\\` for all paths on windows. However, I didn't test the `--cookie` parameter, just the path to `ppx.exe`. Is there still a problem with the `--cookie` path generation? (That's dune).

Summary/next steps:
The next step is to use `create_process`, however, using `create_process` means that we won't get the system shell's automatics/smart parsing of the ppx command string into multiple arguments, and Merlin generates strings in the .merlin file that it assumes will be interpreted by the shell. So the escaping is different in these cases, and also I believe there are some cases where we were relying on the shell to distinguish between spaces in file paths vs. separate exe arguments and to fix this:
1. Merlin would have to then recreate that same shell behavior when interpreting the string.
2. We need to have dune output a different format that requires no interpretation and can just be passed to `create_process` (that was @bryphe's proposal above).

Number 2 requires staging carefully and might require a long process for rolling out these in the right order.
Is number 1 hard or impossible?

**Comment by jordwalke** [comment]
As a super terrible hack, we could also just switch to `create_process` when the config is of the form:

FLG -ppx "path/to/ppx.exe --as-ppx "

When we quickly detect that there are no escaped quotes/backslashes.  There are some projects that don't have the `--cookie` or complicated arguments, and it would at least be nice if the command window didn't pop up for them. Also, for those projects that do have escaping issues right now, they are on windows, and they have to deal with a command window popping up on every keystroke (which means basically unsable anyways). Should we be concerned with breaking users in this ppx workflow that already is not really usable - the benefit is that we can at least make some workflows with ppx work by using `create_process`, whereas before no workflows with ppx would be usable. Thoughts?

**Comment by jordwalke** [comment]
Update: @bryphe I tried that snippet in merlin to use create_process and while it doesn't pop up the dialog, I think it doesn't run the ppx inside of merlin. Did I miss something?

**Comment by jordwalke** [comment]
Update: The issue is that apparently, `Sys.command` escapes things differently than `create_process`:
(Or perhaps the difference comes from `cmd.exe '/c'`:

```reason
let cmd = "echo \"hello!\"";
Sys.command(cmd);
let run_create_process_cmd = (cmd) => {
  let pid =
    Unix.create_process(
      "cmd.exe",
      [|"/c", cmd|],
      Unix.stdin,
      Unix.stderr,
      Unix.stderr,
    );

  let (_, exit) = Unix.waitpid([], pid);
  switch (exit) {
  | Unix.WEXITED(x) => x
  | Unix.WSIGNALED(x) => x
  | Unix.WSTOPPED(x) => x
  };
};
run_create_process_cmd(pwd, cmd);
```

The difference in output:

```
"hello!"
\"hello!\"
```

**Comment by nojb** [comment]
I didn't follow all the details so sorry if this is a bit clueless, but did you try using `Unix.open_process`+ `Filename.quote_command` ?

**Comment by bryphe** [comment]
> Update: The issue is that apparently, Sys.command escapes things differently than create_process:

Hmm, interesting. The `Sys.command` API in OCaml seems to be backed by the [`_wsystem`](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/system-wsystem?view=vs-2019) windows API. It seems like it tries to invoke via `cmd.exe` - but not sure the exact set of arguments it uses (I'm not sure if it uses `/c` for example) - so could be some different behavior.

> I didn't follow all the details so sorry if this is a bit clueless, but did you try using Unix.open_process+ Filename.quote_command ?

@nojb - I think we actually have the reverse problem. If I understand correctly - the `Filename.quote_command` helps if we have a list of arguments, and we want to wrap them in quotes. We have the reverse situation though - we have a long PPX command block - a single string - that we want to _split_ into arguments.

For a command like this: 
```
E:\merlin-language-server demo\_build/default/.ppx//ppx.exe --as-ppx --cookie \"library-name=\\\"Server\\\"\""
```

We'd want to split it into arguments like:
```
["E:\merlin-language-server demo\_build/default/.ppx//ppx.exe", "-as-ppx", "--cookie", "library-name=Server"]
```
(Actually... not 100% sure how the `--cookie` parameter should be passed. But something like that!)

We rely on the shell for this (using `cmd.exe`) - with the undesirable property that it pops open a console window as we're working.

If we could accurately split the _executable_ from the list of args - we could just pass the executable + args to `Unix.create_process` and be good to go (in other words - we could run the process directly without needing the shell `cmd` as an intermediary to parse the command string for us). Hope that helps describe the problem! 

Either `Unix.create_process` or `Unix.open_process` should be OK, since as I can tell they both hit the common codepath that uses the `SW_HIDE` flag to create the process: <https://github.com/ocaml/ocaml/blob/e973cde5a479af088b31de7a1f295256242d3fac/otherlibs/win32unix/createprocess.c#L64>

> Can we use create_process with cmd.exe like your example, but without triggering the quote escaping?

@jordwalke  I think this would work - we'd get the window hidden via the `SW_HIDE` flag, but still leverage `cmd.exe` to do the 'parsing' for us. Do you think we could use a regex or something to sanitize the escaped quotes prior to sending it to create_process / cmd.exe?

**Comment by nojb** [comment]
> @nojb - I think we actually have the reverse problem. If I understand correctly - the `Filename.quote_command` helps if we have a list of arguments, and we want to wrap them in quotes. We have the reverse situation though - we have a long PPX command block - a single string - that we want to _split_ into arguments.

Got it, thanks. One possibility would be to write a `merlin`-specific binding for `CreateProcess` instead of going via `Unix.create_process`. That way, you avoid the need to split the command line at all.

**Comment by let-def** [comment]
`Std.Shell.split_command` (https://github.com/ocaml/merlin/blob/9d70666fd8a4fc454dedabb66410cf9a4c573fbf/src/utils/std.ml#L686) should have the right behavior. A custom binding is fine too.

**Comment by jordwalke** [comment]
I think I found one solution! Check it out. It works well for me on windows, at least in my project with ppx plugins.

**Comment by jordwalke** [comment]
@bryphe Your recent description of the problem does describe my understanding as well. We want the reverse of `quote_command`. The only question was why the escaping was different. I looked at the original PR for `quote_command` and I was kind of mortified to learn how extremely complex it is to parse windows shell commands.

I see my PR as a stop gap that fulfills the contract - PPX in a merlin file instructs a shell. So to guarantee that contract is upheld we run a shell. However, I think there should be a new entry in merlin - which is first supported, and then secondly generated by dune that is more structured.

**Comment by jordwalke** [comment]
@let-def Perhaps `split_command` would work, but I don't see anything in there that is windows specific - wouldn't it have to be to be accurate?

Instead of PPX entries in `.merlin` assuming a shell, can we explicitly split out the exe from the arguments with minimal escaping?

**Comment by let-def** [comment]
@jordwalke The problem is splitting a string into a list of arguments. `Shell.split_command` does that while trying to follow the unix shell specification, without expansion or substitution. So it is more a syntax for specifying a list of strings rather than something system dependent. `.merlin` directives (`-readers`, `-flags`) are already expected to follow shell escaping conventions.

`.merlin` does not assume a shell, just the syntax of a shell.

**Comment by bryphe** [comment]
I was just testing this a bit today (I put together a small repro here: https://github.com/bryphe/merlin-popup-repro). 

And I can observe the 'flashing windows':
![pop-up](https://user-images.githubusercontent.com/13532591/79930207-c3e84d80-83fc-11ea-862a-0cd4db1af1bc.gif)

Now, though, I'm not seeing https://github.com/ocaml/merlin/issues/1034 - the PPX is actually now _running successfully_ - it's just the pesky window popping up that is problematic.

I tested out your PR #1101 @jordwalke - I don't see the 'flashing windows' reproduce, but unfortunately it seems there was an error when running the ppx command:
```
29 items in path, 29 after deduplication
# 27.11 Pparse - apply_rewriter
running "D:\\merlin-language-server\\_build/default/.ppx/dd492af6ac3d011a895e93e95aa5420a/ppx.exe --as-ppx --cookie \"library-name=\\\"Protocol\\\"\"" from directory "D:\\merlin-language-server\\src\\Protocol"
# 27.16 Pparse - report_error
Error while running external preprocessor. Command line: D:\merlin-language-server\_build/default/.ppx/dd492af6ac3d011a895e93e95aa5420a/ppx.exe --as-ppx --cookie "library-name=\"Protocol\"" %merlin_fn_in% %merlin_fn_out% 1>&2
```
(looks like something went wrong with the variable expansion?)

In addition, as another data point, this issue __does not reproduce with `ocaml-lsp'__:

(https://github.com/ocaml/ocaml-lsp):
![ocaml-lsp](https://user-images.githubusercontent.com/13532591/79930482-74eee800-83fd-11ea-9e85-93d805e558a6.gif)

(You can see I'm in a PPX-ed file and getting completion/diagnostics/etc w/o any popup windows). There must be some difference in the way the process is being spawned in the `ocaml-lsp` case (reusing the console, or using `SW_HIDE`)....

Given that, though, I wonder - perhaps it's worth seeing if `ocaml-lsp` addresses this for anyone else?  If that works, maybe it isn't worth investing more in this. But the potential fix I'd see at this point - given that the PPX driver execution is working now - is what @dra27 mentioned above, to use a `CreateProcess` call with the `SW_HIDE` option like @dra27  mentioned above. 

@jordwalke - did this bug still reproduce for you in your gvim setup, with `master` merlin? If it does - might be worth trying [`ocaml-lsp`](https://github.com/ocaml/ocaml-lsp). Curious if that works for you

It's surprising to me that this works now, but a few variables have changed since I tested before (ocaml version, dune version, merlin version, etc...).

**Comment by jordwalke** [comment]
Everything is working fine for me in gvim with merlin and this PR - for my project that uses a ppx. Is there a project that you think I would still experience the flash with? I could try that out.

**Comment by bryphe** [comment]
@jordwalke - interesting - it might be worth running [`procmon`](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon) and [process creation](https://www.winhelponline.com/blog/procmon-track-process-creation-exit-time/) - I wonder what the calls look like for the ppx driver in your case?

I do have a minimal repro:
- `git clone https://github.com/bryphe/merlin-popup-repro`
- `esy`
- `esy x TestPpxParsing.exe`

You should see something like this:

**Comment by jordwalke** [comment]
I found that it was the --cookie param causing the problem.

**Comment by jordwalke** [comment]
@let-def I found that Shell.split_command does will interpret backslashes in file paths incorrectly.

Or you could say it interprets them correctly - assuming a unix shell syntax. However, Dune does not currently generate .merlin files with properly escaped `FLG -ppx "unix shell syntax"` file paths.

Here's an example of `split_command` in action on the contents of a `FLG` parameter in a Dune generated .merlin file:

https://sketch.sh/s/LcheXrEHN3EUvDzpXoKaav/

**Comment by jordwalke** [comment]
I don't think this is so easy to solve under the assumption that merlin FLGs are specified in merlin shell syntax.
I think the best compromise is to have Dune emit .merlin files that use unix shell syntax for flags, where file paths are "as unix as possible" on windows. That means it would need to emit:

```
FLG -ppx "C:/always/forward/slashes/and/if/you/need/a/space/escape\\ it/with/backslashes restofcmd"
```

The only thing I don't think this solves is if your file path has even weirder elements in it such as file paths with backslashes in them.

Then there's the problem of making merlin work on older dunes before this fix is out there in Dune. For that I might try to adjust my PR above to perform some hacks to detect the common case of:

```
FLG -ppx "path/to/ppx.exe anything-else-here"
```

This would only be a stop-gap solution until Dune emits file paths that are as unixy as possible in its ppx command.
are there other common forms I would need to cover?

**Comment by jordwalke** [comment]
@let-def @bryphe I think I might have found a solution in [my PR](https://github.com/ocaml/merlin/pull/1101).  I tested in the case where it previously failed and incorporated the feedback from this discussion.

The problem: Dune outputs paths in the .merlin that are not quite unixy but almost are. Namely, the path to the ppx.exe (which is the first argument). Merlin supplies the paths to the input and output so it guarantees the paths to be formatted as windows paths `C:/`.

So the `FLG` generated by Dune can _almost_ be parsed with `Shell.split_command` but not quite - because of the format of the first path to the executable that Dune generates. So that means I had to split out that path to the ppx executable in the `FLG`, then parse the rest using `Shell.split_command`. But I could not use create_process directly - (and I found that it just invokes cmd.exe as I would anyways [see here](https://github.com/ocaml/ocaml/blob/914dd057b5beb7f0246b8011472105d5096d8783/otherlibs/win32unix/unix.ml#L218)).

I think it has some issues with trying to escape the arguments you pass it in the create_process array, and it might have a bug in escaping (I see commits in the OCaml compiler repo as early as this month trying to fix escaping of commands for windows - so maybe it's still an issue).

So my trick is that I pass these unescaped arguments as strings inside of env variables, once I've split them with `split_command`. It works for my more advanced project which uses `cookie` parameters etc whereas my previous attempts did not. @bryphe do you want to help me test this?

**Comment by voodoos** [comment]
Should be fixed with the latest release (4.2 / 3.5.0). Please reopen if the issue persists.
